src/main/java/com/gargoylesoftware/htmlunit/WebClient.java

```diff
@@ -1681,6 +1679,106 @@ else if (status == HttpStatus.SC_TEMPORARY_REDIRECT
+        }
+
+        // implementation based on rfc9111 https://www.rfc-editor.org/rfc/rfc9111#name-validation
+        if (HeaderUtils.containsETag(cached)) {
+            webRequest.setAdditionalHeader(HttpHeader.IF_NONE_MATCH, cached.getResponseHeaderValue(HttpHeader.ETAG));
+        }
+        if (HeaderUtils.containsLastModified(cached)) {
+            webRequest.setAdditionalHeader(HttpHeader.IF_MODIFIED_SINCE, cached.getResponseHeaderValue(HttpHeader.LAST_MODIFIED));
+        }
+
+        final WebResponse webResponse = getWebConnection().getResponse(webRequest);
+
+        if (webResponse.getStatusCode() >= HttpStatus.SC_INTERNAL_SERVER_ERROR) {
+            return new WebResponseFromCache(cached, webRequest);
+        }
+
+        if (webResponse.getStatusCode() == HttpStatus.SC_NOT_MODIFIED) {
+            final Map<String, NameValuePair> header2NameValuePair = new LinkedHashMap<>();
+            for (NameValuePair pair : cached.getResponseHeaders()) {
+                header2NameValuePair.put(pair.getName(), pair);
+            }
+            for (NameValuePair pair : webResponse.getResponseHeaders()) {
+                if (!shouldDiscardHeaderFrom304Response(pair.getName())) {
+                    header2NameValuePair.put(pair.getName(), pair);
+                }
+            }
+            // WebResponse headers is unmodifiableList so we cannot update it directly
+            // instead, create a new WebResponseFromCache with updated headers then use it to replace the old cached value
+            WebResponse updatedCached = new WebResponseFromCache(cached, new ArrayList<>(header2NameValuePair.values()), webRequest);
+            getCache().cacheIfPossible(webRequest, updatedCached, null);
+            return updatedCached;
+        }
+
+        getCache().cacheIfPossible(webRequest, webResponse, null);
+        return webResponse;
+    }
+
+    /**
+     * These response headers are not copied from a 304 response to the cached
+     * response headers. This list is based on Chromium http_response_headers.cc
+     */
+    private static final String[] DISCARDING_304_RESPONSE_HEADER_NAMES = {
+        "connection",
+        "proxy-connection",
+        "keep-alive",
+        "www-authenticate",
+        "proxy-authenticate",
+        "proxy-authorization",
+        "te",
+        "trailer",
+        "transfer-encoding",
+        "upgrade",
+        "content-location",
+        "content-md5",
+        "etag",
+        "content-encoding",
+        "content-range",
+        "content-type",
+        "content-length",
+        "x-frame-options",
+        "x-xss-protection",
+    };
+
+    private static final String[] DISCARDING_304_HEADER_PREFIXES = {
+        "x-content-",
+        "x-webkit-"
+    };
+
+    /**
+     * Check if this header received from 304 (Not Modified) response should be ignored when updating cached response headers
+     * @param name the header we want to check
```

Comment:
* summary sentence should have period
* avoid using first person like "we"

```suggestion
     * Returns true if the specified header in a 304 Not Modified response should not be used as part of the response.
     * @param name the header to check
```